src/MixedQuoter.sol

```diff
@@ -281,6 +283,159 @@ contract MixedQuoter is IMixedQuoter, IPancakeV3SwapCallback {
+                (swapAmountOut, gasEstimateForCurAction) = quoteExactInputSingleV2(
+                    QuoteExactInputSingleV2Params({tokenIn: tokenIn, tokenOut: tokenOut, amountIn: amountIn})
+                );
+                MixedQuoterTokenRecorder.setPoolSwapTokenAccumulation(poolHash, amountIn, swapAmountOut, isZeroForOne);
+                amountIn = swapAmountOut - accAmountOut;
+            } else if (action == MixedQuoterActions.V3_EXACT_INPUT_SINGLE) {
+                (tokenIn, tokenOut) = convertNativeToWETH(tokenIn, tokenOut);
+                // params[actionIndex]: abi.encode(fee)
+                uint24 fee = abi.decode(params[actionIndex], (uint24));
+
+                bool isZeroForOne = swapDirection(tokenIn, tokenOut);
+                bytes32 poolHash = MixedQuoterTokenRecorder.getV3PoolHash(tokenIn, tokenOut, fee);
+                // update v3 pool swap direction, only allow one direction in one transaction
+                MixedQuoterTokenRecorder.setAndCheckSwapDirection(poolHash, isZeroForOne ? 1 : 2);
+                (uint256 accAmountIn, uint256 accAmountOut) =
+                    MixedQuoterTokenRecorder.getPoolSwapTokenAccumulation(poolHash, isZeroForOne);
+
+                uint256 swapAmountOut;
+                amountIn += accAmountIn;
+                (swapAmountOut,,, gasEstimateForCurAction) = quoteExactInputSingleV3(
+                    QuoteExactInputSingleV3Params({
+                        tokenIn: tokenIn,
+                        tokenOut: tokenOut,
+                        amountIn: amountIn,
+                        fee: fee,
+                        sqrtPriceLimitX96: 0
+                    })
+                );
+                MixedQuoterTokenRecorder.setPoolSwapTokenAccumulation(poolHash, amountIn, swapAmountOut, isZeroForOne);
+                amountIn = swapAmountOut - accAmountOut;
+            } else if (action == MixedQuoterActions.V4_CL_EXACT_INPUT_SINGLE) {
+                QuoteMixedV4ExactInputSingleParams memory clParams =
+                    abi.decode(params[actionIndex], (QuoteMixedV4ExactInputSingleParams));
+                (tokenIn, tokenOut) = convertWETHToV4NativeCurency(clParams.poolKey, tokenIn, tokenOut);
+                bool zeroForOne = tokenIn < tokenOut;
+                checkV4PoolKeyCurrency(clParams.poolKey, zeroForOne, tokenIn, tokenOut);
+
+                bytes32 poolHash = MixedQuoterTokenRecorder.getV4CLPoolHash(clParams.poolKey);
+                // update v4 pool swap direction, only allow one direction in one transaction
+                MixedQuoterTokenRecorder.setAndCheckSwapDirection(poolHash, zeroForOne ? 1 : 2);
+                (uint256 accAmountIn, uint256 accAmountOut) =
+                    MixedQuoterTokenRecorder.getPoolSwapTokenAccumulation(poolHash, zeroForOne);
+
+                uint256 swapAmountOut;
+                amountIn += accAmountIn;
+                (swapAmountOut, gasEstimateForCurAction) = clQuoter.quoteExactInputSingle(
```

Comment:
i think for v4 we need to execute the swap since there might be some side effects from hooks call

Reply:
working on this